crates/fluvio/src/producer/dispatcher.rs

```diff
@@ -0,0 +1,305 @@
+    shutdown: StickyEvent,
+}
+
+enum Event {
+    Command(DispatcherMsg),
+    Timeout,
+}
+
+impl Dispatcher {
+    pub(crate) fn new(
+        spus: Arc<SpuPool>,
+        broadcast_status: tokio::sync::broadcast::Sender<BatchStatus>,
+        commands: Receiver<DispatcherMsg>,
+        config: Arc<ProducerConfig>,
+    ) -> Self {
+        let shutdown = StickyEvent::new();
+        let partition_pool = PartitionPool::new(spus, config.clone(), broadcast_status.clone());
+
+        Self {
+            config,
+            partition_pool,
+            commands: Some(commands),
+            shutdown,
+        }
+    }
+
+    /// Spawn this dispatcher as a task, returning a shutdown handle.
+    pub(crate) fn start(self) -> StickyEvent {
+        let shutdown = self.shutdown.clone();
+        fluvio_future::task::spawn(self.run());
+        shutdown
+    }
+
+    /// Drive this dispatcher via an event loop that reacts to events.
+    pub(crate) async fn run(mut self) {
+        // Set up the event stream: Select between new records and timeouts
+        let mut events = {
+            let timer = async_timer::interval(self.config.batch_duration).map(|_| Event::Timeout);
+            let incoming = self.commands.take().unwrap();
+            let records = incoming.into_stream().map(Event::Command);
+            let select = futures_util::stream::select(timer, records);
+            select.take_until(self.shutdown.listen_pinned())
+        };
+
+        while let Some(event) = events.next().await {
+            let result = self.handle_event(event).await;
+            if let Err(e) = result {
+                error!("Producer Dispatcher error: {}", e);
+            }
+        }
+
+        debug!("Producer dispatcher shutting down");
+    }
+
+    #[instrument(name = "producer_dispatcher", level = "debug", skip(self, event))]
+    async fn handle_event(&mut self, event: Event) -> Result<(), FluvioError> {
+        match event {
+            Event::Command(DispatcherMsg::Record {
+                record,
+                respond_to_producer,
+            }) => {
+                self.partition_pool
+                    .buffer(record, respond_to_producer)
+                    .await?;
+            }
+            Event::Command(DispatcherMsg::Flush(respond_to_producer)) => {
+                let span = debug_span!("flush", reason = "user_flush");
+                self.partition_pool.flush_all().instrument(span).await?;
+
+                // Dropping the sender causes rx.recv() to yield with Err.
+                // This itself is used as the message, and ensures exactly one notification
+                drop(respond_to_producer);
+            }
+            Event::Timeout => {
+                let span = debug_span!("flush", reason = "timeout");
+                self.partition_pool.flush_all().instrument(span).await?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub(crate) struct PartitionPool {
```

Comment:
So is this a shared instance? This would be problematic.  Why not just for this producer instance?

Reply:
Right now it actually is not shared, there is one Dispatcher/PartitionPool per TopicProducer (perhaps we need a different name than PartitionPool). I think we can make it shared at some point in the future